### PR TITLE
PP-4759 Improve international phone number support

### PR DESCRIPTION
**Pull request by DanailMinchev**
## WHAT

This PR adds support for international telephone numbers.

The main changes are:
- new utility for validating and formatting the telephone number `TelephoneNumberUtility`
- changes in invites and users endpoints to support the new validation
- changes in invites and users to format and save the telephone number in `E164` format

with @alexbishop1

### CODE REVIEW

File: src/test/java/uk/gov/pay/adminusers/resources/UserRequestValidatorTest.java [hunk 1]
```diff
@@ -156,22 +156,30 @@ public void shouldSuccess_forSessionVersion_whenPatching() {
...
+    public void shouldSuccess_replacingTelephoneNumber_whenPatchingInternationalTelephoneNumber() {
+        JsonNode payload = new ObjectMapper().valueToTree(ImmutableMap.of("op", "replace", "path", "telephone_number", "value", "+441134960000"));
+        Optional<Errors> optionalErrors = validator.validatePatchRequest(payload);
+
+        assertFalse(optionalErrors.isPresent());
+    }
+
+    @Test
+    public void shouldError_replacingTelephoneNumber_whenPatchingInvalidTelephoneNumber() {
+        JsonNode payload = new ObjectMapper().valueToTree(ImmutableMap.of("op", "replace", "path", "telephone_number", "value", "(╯°□°）╯︵ ┻━┻"));
```
Comment (stephencdaly) [inline src/test/java/uk/gov/pay/adminusers/resources/UserRequestValidatorTest.java]: nice

File: src/test/java/uk/gov/pay/adminusers/utils/telephonenumber/TelephoneNumberUtilityFormatToE164ValidDataTest.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    public void formatToE164_shouldEvaluateToE164FormattedAmericanTelephoneNumber() {
+        // Given
+        String telephoneNumber = "+13115552368";
+        String testResult = "+13115552368";
+
+        // When
+        String result = TelephoneNumberUtility.formatToE164(telephoneNumber);
+
+        // Then
+        assertThat("Expected " + telephoneNumber + " to be " + testResult, result, is(testResult));
```
Comment (stephencdaly) [inline src/test/java/uk/gov/pay/adminusers/utils/telephonenumber/TelephoneNumberUtilityFormatToE164ValidDataTest.java]: providing a message as an argument is probably unnecessary here, as a similar message will be constructed by default

File: src/test/java/uk/gov/pay/adminusers/utils/telephonenumber/TelephoneNumberUtilityFormatToE164InvalidDataTest.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    @Parameterized.Parameters
+    public static Object[] data() {
+        return new Object[]{
+                null,
+                "",
+                " ",
+                "abc",
+                "(╯°□°）╯︵ ┻━┻"
+        };
+    }
```
Comment (stephencdaly) [inline src/test/java/uk/gov/pay/adminusers/utils/telephonenumber/TelephoneNumberUtilityFormatToE164InvalidDataTest.java]: nice use of parameterized tests

File: src/test/java/uk/gov/pay/adminusers/utils/telephonenumber/TelephoneNumberUtilityFormatToE164ValidUnitedKingdomDataTest.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                "(0044113) 496-0000",
+                "(0044113) / 496-0000",
+                "   00441134960000   "
+        };
+    }
+
+    @Test
+    public void formatToE164_shouldEvaluateToE164FormattedTelephoneNumber() {
+        String result = TelephoneNumberUtility.formatToE164(telephoneNumber);
+        assertThat("Expected " + telephoneNumber + " to be " + TEST_RESULT, result, is(TEST_RESULT));
```
Comment (stephencdaly) [inline src/test/java/uk/gov/pay/adminusers/utils/telephonenumber/TelephoneNumberUtilityFormatToE164ValidUnitedKingdomDataTest.java]: same here, message argument probably unnecessary

File: src/test/java/uk/gov/pay/adminusers/utils/telephonenumber/TelephoneNumberUtilityIsValidPhoneNumberTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                {"0770090000a", false},
+                {"O2O793O0000", false}, // letter "O" instead of "0"
+                {"abc", false}
+        });
+    }
+
+    @Test
+    public void isValidPhoneNumber_shouldEvaluateWhetherOrNotItIsValidPhoneNumber() {
+        boolean result = TelephoneNumberUtility.isValidPhoneNumber(telephoneNumber);
+        assertThat("Expected " + telephoneNumber + " to be " + (testResult ? "valid" : "invalid"), result, is(testResult));
```
Comment (stephencdaly) [inline src/test/java/uk/gov/pay/adminusers/utils/telephonenumber/TelephoneNumberUtilityIsValidPhoneNumberTest.java]: here the message argument does make sense, as it makes it clearer what the assertion is, nice 👍

File: src/main/java/uk/gov/pay/adminusers/service/NotificationService.java [hunk 6]
```diff
@@ -61,11 +61,18 @@ NotifyDirectDebitConfiguration getNotifyDirectDebitConfiguration() {
 
     CompletableFuture<String> sendSecondFactorPasscodeSms(String phoneNumber, String passcode) {
         return CompletableFuture.supplyAsync(() -> {
+            // convert back to local format for backward compatibility
+            String formattedPhoneNumber;
+            if (phoneNumber.startsWith("+44")) {
+                formattedPhoneNumber = phoneNumber.replaceFirst("\\+44", "0");
+            } else {
+                formattedPhoneNumber = phoneNumber;
+            }
```
Comment (stephencdaly) [inline src/main/java/uk/gov/pay/adminusers/service/NotificationService.java]: Is this change going to be reverted after the end-to-end tests have been updated?

Comment (DanailMinchev) [reply src/main/java/uk/gov/pay/adminusers/service/NotificationService.java]: Yes, it will be reverted once we migrate fully, thank you for asking

**Pull request merged**